src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Selectors/SamlSecurityTokenAuthenticator.cs

```diff
@@ -0,0 +1,243 @@
+            }
+
+            _maxClockSkew = maxClockSkew;
+            _audienceUriMode = AudienceUriMode.Always;
+            _allowedAudienceUris = new Collection<string>();
+        }
+
+        public AudienceUriMode AudienceUriMode
+        {
+            get { return _audienceUriMode; }
+            set
+            {
+                AudienceUriModeValidationHelper.Validate(_audienceUriMode);
+                _audienceUriMode = value;
+            }
+        }
+
+        public IList<string> AllowedAudienceUris
+        {
+            get { return _allowedAudienceUris; }
+        }
+
+        protected override bool CanValidateTokenCore(SecurityToken token)
+        {
+            return token is SamlSecurityToken;
+        }
+
+        protected override ValueTask<ReadOnlyCollection<IAuthorizationPolicy>> ValidateTokenCoreAsync(Tokens.SecurityToken token)
+        {
+            if (token == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(token));
+
+            SamlSecurityToken samlToken = token as SamlSecurityToken;
+
+            if (samlToken == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SamlTokenAuthenticatorCanOnlyProcessSamlTokens, token.GetType().ToString())));
+
+            if (samlToken.Assertion.Signature == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SamlTokenMissingSignature)));
+
+            if (!IsCurrentlyTimeEffective(samlToken))
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SAMLTokenTimeInvalid, DateTime.UtcNow.ToUniversalTime(), samlToken.ValidFrom.ToString(), samlToken.ValidTo.ToString())));
+
+          //  if (samlToken.Assertion.SigningToken == null)
+          //      throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SamlSigningTokenMissing)));
+
+            // Build the Issuer ClaimSet for this Saml token.
+            ClaimSet issuer = null;
+            bool canBeValidated = false;
+            for (int i = 0; i < _supportingAuthenticators.Count; ++i)
+            {
+                canBeValidated = _supportingAuthenticators[i].CanValidateToken(samlToken);
+                if (canBeValidated)
+                    break;
+            }
+            if (!canBeValidated)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SamlInvalidSigningToken)));
+            ReadOnlyCollection<IAuthorizationPolicy> policies;
+            issuer = ResolveClaimSet(samlToken, out policies) ?? ClaimSet.Anonymous;
+
+            if ((_audienceUriMode == AudienceUriMode.Always)
+             || (_audienceUriMode == AudienceUriMode.BearerKeyOnly) && (samlToken.SecurityKeys.Count < 1))
```

Comment:
You dropped some code here. There should be a loop which iterates over samlToken.Assertion.Statements and calls CreatePolicy on each one. I know you modified ResolveClaimsSet to return the authorization policies, but they are going to be or can be different.